SimPEG/Regularization.py

```diff
@@ -416,655 +575,802 @@ def eval2Deriv(self, m, v=None):
-            else:
-                m = self.mapping * (self.model)
 
-            mats = []
-            for imodel in range(self.nModels):
+class Sparse(BaseComboRegularization):
+    """
+    The regularization is:
 
-                indl, indu = imodel*self.regmesh.nC, (imodel+1)*self.regmesh.nC
+    .. math::
 
-                # Grab the right model parameters
-                f_m = self.regmesh.cellDiffxStencil * m[indl:indu]
-                self.rx = self.R( f_m , self.eps_q[imodel], self.norms[1])
+        R(m) = \\frac{1}{2}\mathbf{(m-m_\\text{ref})^\\top W^\\top R^\\top R
+        W(m-m_\\text{ref})}
 
-                Wx = Utils.sdiag((self.alpha_x*self.gamma*(self.regmesh.aveCC2Fx*self.cell_weights[indl:indu]))**0.5*self.rx)*self.regmesh.cellDiffxStencil
+    where the IRLS weight
 
-                mats.append(Wx)
+    .. math::
 
-            self._Wx = sp.block_diag(mats)
+        R = \eta TO FINISH LATER!!!
 
-        return self._Wx
+    So the derivative is straight forward:
 
-    @property
-    def Wy(self):
-        """Regularization matrix Wy"""
-        if getattr(self, '_Wy', None) is None:
+    .. math::
 
-            if getattr(self, 'model', None) is None:
-                m = np.ones(self.mapping.shape[0])
+        R(m) = \mathbf{W^\\top R^\\top R W (m-m_\\text{ref})}
 
-            else:
-                m = self.mapping * (self.model)
+    The IRLS weights are recomputed after each beta solves.
+    It is strongly recommended to do a few Gauss-Newton iterations
+    before updating.
+    """
+    def __init__(
+        self, mesh,
+        alpha_s=1.0, alpha_x=1.0, alpha_y=1.0, alpha_z=1.0,
+        **kwargs
+    ):
 
-            mats = []
-            for imodel in range(self.nModels):
+        objfcts = [
+            SparseSmall(mesh=mesh, **kwargs),
+            SparseDeriv(mesh=mesh, orientation='x', **kwargs)
+        ]
 
-                indl, indu = imodel*self.regmesh.nC, (imodel+1)*self.regmesh.nC
+        if mesh.dim > 1:
+            objfcts.append(SparseDeriv(mesh=mesh, orientation='y', **kwargs))
 
-                # Grab the right model parameters
-                f_m = self.regmesh.cellDiffyStencil * m[indl:indu]
-                self.ry = self.R( f_m , self.eps_q[imodel], self.norms[2])
+        if mesh.dim > 2:
+            objfcts.append(SparseDeriv(mesh=mesh, orientation='z', **kwargs))
 
-                Wy = Utils.sdiag((self.alpha_y*self.gamma*(self.regmesh.aveCC2Fy*self.cell_weights[indl:indu]))**0.5*self.ry)*self.regmesh.cellDiffyStencil
+        super(Sparse, self).__init__(
+            mesh=mesh, objfcts=objfcts,
+            alpha_s=alpha_s, alpha_x=alpha_x, alpha_y=alpha_y, alpha_z=alpha_z,
+            **kwargs
+        )
 
-                mats.append(Wy)
+        # Utils.setKwargs(self, **kwargs)
 
-            self._Wy = sp.block_diag(mats)
+    # Properties
+    norms = properties.Array(
+        "Norms used to create the sparse regularization",
+        default=[0., 2., 2., 2.]
```

Comment:
@fourndo: Do we want the default to be `[0., 2., 2., 2.]` or should it be `[2., 2., 2., 2.]`?